### PR TITLE
Typescript for createReanimatedComponent

### PR DESCRIPTION
**Pull request by jmysliv**
## Description
As the next step of the react-native-reanimated smooth transition to TypeScript, this PR updates types for all remaining js-modules.

## Changes
* updates types for `createReanimatedComponent`
* update types for `js-reanimated`
* update types for `ConfigHelper`
* update types for `setAndForwardRef`

### CODE REVIEW

File: src/reanimated2/js-reanimated/MapperRegistry.ts [hunk 1]
```diff
@@ -35,10 +37,10 @@ export default class MapperRegistry {
     }
   }
 
-  updateOrder() {
+  updateOrder(): void {
     const nodes = [...this.mappers.values()].map((mapper) => new Node(mapper));
 
-    const mappersById = {};
+    const mappersById: { [key: number]: Mapper<T> } = {};
```
Comment (jmysliv) [inline src/reanimated2/js-reanimated/MapperRegistry.ts:43]: Change to record

File: src/reanimated2/js-reanimated/MapperRegistry.ts [hunk 2]
```diff
@@ -75,9 +77,9 @@ export default class MapperRegistry {
       }
     }
 
-    const post = {};
+    const post: { [key: number]: number } = {};
```
Comment (jmysliv) [inline src/reanimated2/js-reanimated/MapperRegistry.ts:80]: change to record

File: src/ConfigHelper.ts [hunk 3]
```diff
@@ -133,17 +135,23 @@ export function addWhitelistedUIProps(props) {
 }
 
 const PROCESSED_VIEW_NAMES = new Set();
+
+interface ViewConfig {
+  uiViewClassName: string;
+  validAttributes: Record<string, any>;
```
Comment (jakub-gonet) [inline src/ConfigHelper.ts:141]: What `validAttributes` represent?

File: src/ReanimatedModuleCompat.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    // noop
+  },
+  async detachEvent(
+    _viewTag: number,
+    _eventName: string,
+    _nodeID: number
+  ): Promise<void> {
+    // noop
+  },
+  async createNode(_nodeID: number, _config: any): Promise<void> {
```
Comment (jakub-gonet) [inline src/ReanimatedModuleCompat.ts:19]: Why config is `any` instead of `Record`?

File: src/reanimated2/js-reanimated/index.ts [hunk 5]
```diff
@@ -50,7 +59,7 @@ export const _updatePropsJS = (updates, viewRef) => {
           return;
         }
         const dashedKey = key.replace(/[A-Z]/g, (m) => '-' + m.toLowerCase());
-        viewRef._component._touchableNode.setAttribute(
+        viewRef._component?._touchableNode.setAttribute(
```
Comment (jakub-gonet) [inline src/reanimated2/js-reanimated/index.ts:62]: Why though?

File: src/reanimated2/js-reanimated/index.ts [hunk 6]
```diff
@@ -1,10 +1,16 @@
...
 import { shouldBeUseWeb } from '../PlatformChecker';
+import { AnimatedStyle, StyleProps } from '../commonTypes';
 
 const reanimatedJS = new JSReanimated();
 
+interface JSReanimatedComponent {
+  previousStyle: StyleProps;
+  setNativeProps: (style: StyleProps) => void;
+  props: Record<string, string | number>;
+  _touchableNode: any;
```
Comment (jakub-gonet) [inline src/reanimated2/js-reanimated/index.ts:11]: What is `_touchableNode`?

File: src/createAnimatedComponent.tsx [hunk 7]
```diff
@@ -34,7 +53,7 @@ function dummyListener() {
   // event is used.
 }
 
-function hasAnimatedNodes(value) {
+function hasAnimatedNodes(value: any): boolean {
```
Comment (jmysliv) [inline src/createAnimatedComponent.tsx:56]: change to unknown

File: src/reanimated2/js-reanimated/JSReanimated.ts [hunk 8]
```diff
@@ -68,24 +57,16 @@ export default class JSReanimated extends NativeReanimated {
...
   }
 
   makeShareable<T>(value: T): T {
     return value;
   }
 
-  makeMutable<T>(value: T): MutableValue {
-    return new MutableValue(value, this._valueSetter);
+  makeMutable<T>(value: T): MutableValue<T> {
+    return new MutableValue(value, this._valueSetter as <T>(value: T) => void);
```
Comment (jakub-gonet) [inline src/reanimated2/js-reanimated/JSReanimated.ts:69]: Why you cast it here?

File: src/createAnimatedComponent.tsx (hunk 7 above)
Comment (jakub-gonet) [inline src/createAnimatedComponent.tsx:56]:
```suggestion
function hasAnimatedNodes(value: unknown): boolean {
```

File: src/createAnimatedComponent.tsx [hunk 9]
```diff
@@ -66,21 +86,85 @@ function flattenArray(array) {
...
 
-export default function createAnimatedComponent(Component, options = {}) {
+interface AnimatedProps extends Record<string, unknown> {
+  viewDescriptors?: ViewDescriptorsSet;
+  viewsRef?: ViewRefSet<unknown>;
+  initial?: SharedValue<StyleProps>;
+}
+
+export type AnimatedComponentProps<P extends Record<string, unknown>> = {
+  [K in keyof P]: P[K];
```
Comment (jakub-gonet) [inline src/createAnimatedComponent.tsx:96]: I don't get this type.

File: src/createAnimatedComponent.tsx [hunk 10]
```diff
@@ -66,21 +86,85 @@ function flattenArray(array) {
...
+  getScrollableNode?: () => ComponentRef;
+}
+
+const has = <K extends string>(
+  key: K,
+  x: unknown
+): x is { [key in K]: unknown } => {
+  if (typeof x === 'function' || typeof x === 'object') {
+    if (x === null || x === undefined) return false;
+    else return key in x;
```
Comment (jakub-gonet) [inline src/createAnimatedComponent.tsx:133]: Brackets.

File: src/createAnimatedComponent.tsx [hunk 11]
```diff
@@ -150,16 +234,18 @@ export default function createAnimatedComponent(Component, options = {}) {
...
         for (const style of this._styles) {
           if (style?.viewsRef) {
             style.viewsRef.remove(this);
           }
         }
       } else if (this._viewTag !== -1) {
-        for (const style of this._styles) {
-          if (style?.viewDescriptors) {
-            style.viewDescriptors.remove(this._viewTag);
+        if (this._styles !== null) {
```
Comment (jakub-gonet) [inline src/createAnimatedComponent.tsx:244]: You can merge those ifs.

File: src/createAnimatedComponent.tsx [hunk 12]
```diff
@@ -208,15 +296,18 @@ export default function createAnimatedComponent(Component, options = {}) {
...
-          prop.attachEvent(node, key);
+          (prop as AnimatedEvent).attachEvent(node, key);
         } else if (
-          prop?.current &&
+          has('current', prop) &&
           prop.current instanceof WorkletEventHandler &&
           prop.current.reattachNeeded
         ) {
-          prop.current.registerForEvents(viewTag, key);
+          prop.current.registerForEvents(viewTag as number, key);
```
Comment (jakub-gonet) [inline src/createAnimatedComponent.tsx:310]: Why you have to cast here?

File: src/createAnimatedComponent.tsx [hunk 13]
```diff
@@ -335,13 +427,13 @@ export default function createAnimatedComponent(Component, options = {}) {
       // attach animatedProps property
       if (this.props.animatedProps?.viewDescriptors) {
         this.props.animatedProps.viewDescriptors.add({
-          tag: viewTag,
-          name: viewName,
+          tag: viewTag as number,
+          name: viewName as string,
```
Comment (jakub-gonet) [inline src/createAnimatedComponent.tsx:431]: You don't need cast here. Null assertion will suffice.

File: src/ConfigHelper.ts (hunk 3 above)
Comment (jmysliv) [reply src/ConfigHelper.ts:141]: properties of given `viewConfig` that should be added to props whitelist.

Comment (jakub-gonet) [reply src/ConfigHelper.ts:141]: so `Record<string, boolean>`?

Comment (jmysliv) [reply src/ConfigHelper.ts:141]: I guess there might be other types than `boolean` as well, see: https://github.com/facebook/react-native/blob/main/Libraries/Components/ScrollView/ScrollViewViewConfig.js#L32. 
I think the best option will be `Record<string, unknown>` then.

File: src/createAnimatedComponent.tsx (hunk 9 above)
Comment (jmysliv) [reply src/createAnimatedComponent.tsx:112]: It simply merges props of the given component with props that are used by our AnimatedComponent.

**Review by jakub-gonet — APPROVED** [review]

**Review by piaskowyk — APPROVED** [review]

**Pull request merged**